pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -4278,16 +4287,33 @@ func mapNodeNameToPrivateDNSName(nodeName types.NodeName) string {
 }
 
 // mapInstanceToNodeName maps a EC2 instance to a k8s NodeName, by extracting the PrivateDNSName
-func mapInstanceToNodeName(i *ec2.Instance) types.NodeName {
+func mapInstanceToNodeName(i *ec2.Instance, explicitNames bool) types.NodeName {
+	if explicitNames {
+		for _, tag := range i.Tags {
+			key := aws.StringValue(tag.Key)
+			if key == TagNameKubernetesNodeNamePrefix {
+				value := aws.StringValue(tag.Value)
+				if len(value) > 0 {
+					return types.NodeName(aws.StringValue(tag.Value))
+				}
+				glog.Infof("Tag %s was found for instance %s, but had no value", TagNameKubernetesNodeNamePrefix, aws.StringValue(i.InstanceId))
```

Comment:
will silently falling back to the private dns name lead to scenarios where there's a race to apply the custom tag before the kubelet starts on up on the node and registers itself using the private dns name?

Reply:
Enabling this feature and not having the tags in place will cause problems, but I think if you decide to enable this feature it's on you to make sure you don't mess up the names. This is already the case with the `cloud-provider` flag which requires the node-name and hostname to match (something that is undocumented).

Reply:
Also, my guess is that kubelet doesn't register if this happens.

Reply:
> my guess is that kubelet doesn't register if this happens

if it's the first startup in a tls-bootstrapping setup, the kublet will use the private dns name to get it's initial credential and register, then start failing as soon as the tag appears and it's idea of who it is changes

Reply:
> Enabling this feature and not having the tags in place will cause problems

This must be clearly documented. And IMO kubelet should exit when it starts on a node with ExplicitNodeNames enabled and without the right tag on the node.